Mage/src/main/java/mage/abilities/effects/ReturnMORToBattlefieldUnderOwnerControlWithCounterEffect.java

```diff
@@ -0,0 +1,128 @@
+        if (card != null && objectToReturn.refersTo(card, game)) {
+            Player owner = game.getPlayer(card.getOwnerId());
+            if (owner != null) {
+                if (card instanceof MeldCard) {
+                    MeldCard meldCard = (MeldCard) card;
+                    game.addEffect(new ReturnMOREntersBattlefieldEffect(
+                            new MageObjectReference(meldCard.getTopHalfCard(), game),
+                            counters,
+                            counterText
+                    ), source);
+                    game.addEffect(new ReturnMOREntersBattlefieldEffect(
+                            new MageObjectReference(meldCard.getBottomHalfCard(), game),
+                            counters,
+                            counterText
+                    ), source);
+                } else {
+                    game.addEffect(new ReturnMOREntersBattlefieldEffect(objectToReturn, counters, counterText), source);
+                }
+                owner.moveCards(card, Zone.BATTLEFIELD, source, game, false, false, true, null);
+            }
+        }
+        return true;
+    }
+}
+
+class ReturnMOREntersBattlefieldEffect extends ReplacementEffectImpl {
+
+    private final MageObjectReference objectToReturn;
+    private final Set<Counter> counters;
+
+    public ReturnMOREntersBattlefieldEffect(MageObjectReference objectToReturn, Set<Counter> counters, String counterText) {
+        super(Duration.EndOfStep, Outcome.BoostCreature);
+        this.objectToReturn = objectToReturn;
+        this.counters = new HashSet<>(counters);
+        this.staticText = "that card to the battlefield with " + counterText + " on it";
+    }
+
+    private ReturnMOREntersBattlefieldEffect(final ReturnMOREntersBattlefieldEffect effect) {
+        super(effect);
+        this.objectToReturn = effect.objectToReturn;
+        this.counters = new HashSet<>(effect.counters);
+    }
+
+    @Override
+    public ReturnMOREntersBattlefieldEffect copy() {
+        return new ReturnMOREntersBattlefieldEffect(this);
+    }
+
+    @Override
+    public boolean checksEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.ENTERS_THE_BATTLEFIELD;
+    }
+
+    @Override
+    public boolean applies(GameEvent event, Ability source, Game game) {
+        if (event.getType() == GameEvent.EventType.ENTERS_THE_BATTLEFIELD) {
+            return event.getTargetId().equals(objectToReturn.getSourceId());
+        }
+        return false;
+    }
+
+    @Override
+    public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+        Permanent permanent = ((EntersTheBattlefieldEvent) event).getTarget();
+        if (permanent != null) {
+            counters.forEach(counter ->
+                    permanent.addCounters(counter, source.getControllerId(), source, game, event.getAppliedEffects())
+            );
+            discard(); // use only once
```

Comment:
this is a bit strange to me with the `discard()`. There's another way of doing it with `game.setEnterWithCounters()`, see Cauldron's Gift for an example. Might be simpler (if you think it's not, then it's okay as is)

Reply:
Thanks for pointing that out. Missed that method entirely, and used the two other cards as start for the common effect.